File: 4_semester/CaAA/Python/lb5/main.py
class BohrNode:
    """
    This is the node class of the Bohr structure
    """

    def __init__(self) -> None:
        """
        This method initializes the initial node parameters needed to solve the problem
        """

        self.children_nodes = {}
        self.patterns_indexes = []
        self.suffix_link = None


class Solver:
    """
    This class solves the problem of finding occurrences of a substring in a string
    """

    def __init__(self, task_type: str = "Classic") -> None:
        """
        This method initializes the initial parameters needed to solve the problem
        """

        self.__task_type = task_type
        self.__bohr_root = BohrNode()
        self.__text = str()
        self.__patterns = list()
        self.__answer = list()

        self.__count_patterns = int()
        self.__pattern_parts_indexes = list()
        self.__joker_pattern = str()
        self.__joker_symbol = str()

    def __classic_input(self) -> None:
        """
        This method reads the input data for the classical algorithm
        :return: None
        """

        self.__text = input()
        self.__count_patterns = int(input())
        self.__patterns = [input() for _ in range(self.__count_patterns)]

    def __joker_input(self) -> None:
        """
        This method reads the input data for the algorithm with joker
        :return: None
        """

        self.__text = input()
        self.__joker_pattern = input()
        self.__joker_symbol = input()

    def __format_joker_input(self) -> None:
        """
        This method brings the given to the classical form, for using the classical algorithm
        :return: None
        """

        self.__patterns = list(filter(bool, self.__joker_pattern.split(self.__joker_symbol)))
        self.__count_patterns = len(self.__patterns)
        is_joker_symbol = True

        for index, symbol in enumerate(self.__joker_pattern):
            if symbol == self.__joker_symbol:
                is_joker_symbol = True
                continue
            if is_joker_symbol:
                self.__pattern_parts_indexes.append(index)
                is_joker_symbol = False

    def read_data(self) -> None:
        """
        This method reads given by the desired method, depending on the type of task
        :return: None
        """

        if self.__task_type == "Joker":
            self.__joker_input()
            self.__format_joker_input()
        else:
            self.__classic_input()

    def __build_bohr(self) -> None:
        """
        This method builds a bohr structure
        :return: None
        """

        for pattern_index in range(self.__count_patterns):
            current_node = self.__bohr_root

            for symbol in self.__patterns[pattern_index]:
                current_node = current_node.children_nodes.setdefault(symbol, BohrNode())

            current_node.patterns_indexes.append(pattern_index)

    def __build_suffix_links(self) -> None:
        """
        This method builds suffix links in the Bohr using a breadth-first crawl
        :return: None
        """

        queue = []
        for child_node in self.__bohr_root.children_nodes.values():
            child_node.suffix_link = self.__bohr_root
            queue.append(child_node)

        while len(queue) > 0:
            current_node = queue.pop(0)

            for symbol, child_node in current_node.children_nodes.items():
                queue.append(child_node)
                current_suffix_node = current_node.suffix_link

                while (current_suffix_node is not None) and (symbol not in current_suffix_node.children_nodes.keys()):
                    current_suffix_node = current_suffix_node.suffix_link

                if current_suffix_node:
                    child_node.suffix_link = current_suffix_node.children_nodes[symbol]
                else:
                    child_node.suffix_link = self.__bohr_root
                child_node.patterns_indexes += child_node.suffix_link.patterns_indexes

    def __aho_korasik_algorithm(self) -> None:
        """
        This method solves the problem using the classical Aho-Korasik algorithm
        :return: None
        """

        current_node = self.__bohr_root

        for symbol_index, symbol in enumerate(self.__text):
            while (current_node is not None) and (symbol not in current_node.children_nodes):
                current_node = current_node.suffix_link

            if current_node is None:
                current_node = self.__bohr_root
                continue

            current_node = current_node.children_nodes[symbol]
            for pattern_index in current_node.patterns_indexes:
                self.__answer.append((symbol_index - len(self.__patterns[pattern_index]) + 1, pattern_index))

        self.__answer = sorted(self.__answer)


    def __joker_aho_korasik_algorithm(self) -> None:
        """
        This method solves the problem with joker using the classical Aho-Korasik algorithm
        :return: None
        """

        self.__aho_korasik_algorithm()

        counter = [0] * len(self.__text)
        for aho_index, pattern_index  in self.__answer:
            joker_pattern_occurance_index = aho_index - self.__pattern_parts_indexes[pattern_index]
            if (joker_pattern_occurance_index >= 0) and \
                    (joker_pattern_occurance_index < len(self.__text)):
                counter[joker_pattern_occurance_index] += 1

        joker_answer = list()
        for index in range(len(self.__text) - len(self.__joker_pattern) + 1):
            if counter[index] == len(self.__patterns):
                joker_answer.append(index + 1)
        self.__answer = joker_answer

    def solve(self) -> None:
        """
        This method solves the problem with the right method, depending on the type of task
        :return: None
        """

        self.__build_bohr()
        self.__build_suffix_links()
        if self.__task_type == "Joker":
            self.__joker_aho_korasik_algorithm()
        else:
            self.__aho_korasik_algorithm()

    def print_answer(self) -> None:
        """
        This method outputs the answer in the format that a particular task requires
        :return: None
        """

        if self.__task_type == "Joker":
            print(*self.__answer, sep='\n')
        else:
            for pair in self.__answer:
                print(f"{pair[0] + 1} {pair[1] + 1}")


def main():
    """
    This function solves the problem
    :return: None
    """
    #solver = Solver("Joker")
    solver = Solver()
    solver.read_data()
    solver.solve()
    solver.print_answer()


if __name__ == "__main__":
    main()

File: 4_semester/CaAA/C++/lb5/main.cpp
#include <iostream>
#include <cstring>
#include <vector>
#include <map>
#include <queue>
#include <algorithm>


#define ALPHABET 5
enum TaskTypes {Classic, Joker};

bool cmp(const std::pair<int, int> &a, const std::pair<int, int> &b){
    if (a.first != b.first){
        return a.first < b.first;
    }
    return a.second < b.second;
}


class BohrNode{
    private:
        std::map<char, BohrNode*> children_nodes;
        BohrNode* suffix_link;
        std::vector<int> patterns_indexes;

    public:
        BohrNode();
        std::map<char, BohrNode*> getChildrenNodes();
        BohrNode* getChild(char suffix);
        void setChild(char suffix, BohrNode* child_node);
        std::vector<int> getPatternsIndexes();
        void addPatternIndex(size_t pattern_index);
        BohrNode* getSuffixLink();
        void setSuffixLink(BohrNode* suffix_node);
};

BohrNode::BohrNode(){
    this->suffix_link = nullptr;
};

std::map<char, BohrNode*> BohrNode::getChildrenNodes(){
    return this->children_nodes;
};

BohrNode* BohrNode::getChild(char suffix){
    auto iterator = this->children_nodes.find(suffix);
    if (iterator != this->children_nodes.end()){
        return iterator->second;
    }
    return nullptr;
};

void BohrNode::setChild(char suffix, BohrNode* child_node){
    auto iterator = this->children_nodes.find(suffix);
    if (iterator == this->children_nodes.end()){
        this->children_nodes.insert({suffix, child_node});
    } else {
        iterator->second = child_node;
    }
};

std::vector<int> BohrNode::getPatternsIndexes(){
    return this->patterns_indexes;
};

void BohrNode::addPatternIndex(size_t pattern_index){
    this->patterns_indexes.emplace_back(pattern_index);
};

BohrNode* BohrNode::getSuffixLink(){
    return this->suffix_link;
};

void BohrNode::setSuffixLink(BohrNode* suffix_node){
    this->suffix_link = suffix_node;
};


class Solver{
    private:
        TaskTypes task_type;
        BohrNode* bohr_root;
        std::string text;
        std::vector<std::string> patterns;
        std::vector<std::pair<int, int>> patterns_occurrence_indexes;

        size_t count_patterns;
        std::string joker_pattern;
        char jokerSymbol;
        std::vector<int> indexes_of_pattern_parts;
        std::vector<int> joker_pattern_occurence_indexes;

        void buildBohr();
        void buildSuffixAndOutputLinks();
        void classicInput();
        void jokerInput();
        void formatJokerInputData();
        void solveClassic();
        void solveJoker();
        void printClassic();
        void printJoker();

    public:
        Solver(TaskTypes task_type);
        void solve();
        void printAnswer();
};

Solver::Solver(TaskTypes task_type=Classic){
    this->task_type = task_type;
    
    switch (this->task_type){
        case Classic: 
            this->classicInput();
            break;
        case Joker:
            this->jokerInput();
            this->formatJokerInputData();
            break;
        default:
            this->classicInput();
            break;
    }
    
    this->bohr_root = new BohrNode();
};

void Solver::classicInput(){
    std::cin >> this->text;
    std::cin >> this->count_patterns;
    
    for(size_t i = 0; i < this->count_patterns; i++){
        std::string pattern;
        std::cin >> pattern;
        patterns.push_back(pattern);
    }  
};

void Solver::jokerInput(){
    std::cin >> this->text;
    std::cin >> this->joker_pattern;
    std::cin >> this->jokerSymbol;
};

void Solver::formatJokerInputData(){
    std::string input_pattern = this->joker_pattern + this->jokerSymbol;
    size_t joker_index = 0;
    size_t index_occurance_part = 0;
    while ((joker_index = input_pattern.find(this->jokerSymbol)) != std::string::npos){
        std::string pattern_part = input_pattern.substr(0, joker_index);
        input_pattern.erase(0, joker_index + 1);
        if (pattern_part.size() != 0 && pattern_part != input_pattern){
            this->patterns.push_back(pattern_part);
            this->indexes_of_pattern_parts.push_back(index_occurance_part);
        }
        index_occurance_part += joker_index + 1;
    }
};

void Solver::buildBohr(){
    for (size_t index = 0; index < this->count_patterns; index++){
        BohrNode* current_node = this->bohr_root;

        for (char symbol : patterns.at(index)){
            if (current_node->getChild(symbol) == nullptr){
                current_node->setChild(symbol, new BohrNode());
            }
            current_node = current_node->getChild(symbol);
        }
        current_node->addPatternIndex(index);
    }
};

void Solver::buildSuffixAndOutputLinks(){
    std::queue<BohrNode*> queue;

    for(auto &iterator : this->bohr_root->getChildrenNodes()){
        iterator.second->setSuffixLink(this->bohr_root);
        queue.push(iterator.second);
    }

    while(queue.size() > 0){
        BohrNode* current_node = queue.front();
        queue.pop();

        for(auto &iterator : current_node->getChildrenNodes()){
            char suffix = iterator.first;
            BohrNode* child_node = iterator.second;

            queue.push(child_node);
            BohrNode* suffix_link_node = current_node->getSuffixLink();
            
            while(suffix_link_node != nullptr && suffix_link_node->getChild(suffix) == nullptr){
                suffix_link_node = suffix_link_node->getSuffixLink();
            }

            if(suffix_link_node != nullptr){
                child_node->setSuffixLink(suffix_link_node->getChild(suffix));
            } else {
                child_node->setSuffixLink(this->bohr_root);
            } 
            for (size_t pattern_index : child_node->getSuffixLink()->getPatternsIndexes()){
                child_node->addPatternIndex(pattern_index);
            }
        }
    }
};

void Solver::solveClassic(){
    BohrNode* current_node = this->bohr_root;

    for (int index = 0; index < this->text.length(); index++){
        char symbol = this->text[index];
        
        while (current_node != nullptr && current_node->getChild(symbol) == nullptr){
            current_node = current_node->getSuffixLink();
        }

        if (current_node == nullptr){
            current_node = this->bohr_root;
            continue;
        }

        current_node = current_node->getChild(symbol);
        for (size_t pattern_index : current_node->getPatternsIndexes()){
            int pattern_size =  this->patterns.at(pattern_index).size();
            this->patterns_occurrence_indexes.emplace_back(std::pair<int, int>{index - pattern_size + 1, pattern_index});
        }
    }
    std::sort(this->patterns_occurrence_indexes.begin(), this->patterns_occurrence_indexes.end(), cmp);
};

void Solver::solveJoker(){
    this->solveClassic();
    std::cout<<"occurence part indexes\n";
    for (auto p : this->patterns_occurrence_indexes){
        std::cout<<p.first<<' '<<p.second<<'\n';
    }
    std::cout<<"occurence part indexes\n";
    
    std::vector<int> c(this->text.size());
    for (auto pair : this->patterns_occurrence_indexes){
        int occurance_index = pair.first;
        int pattern_index = pair.second;
        int idx = occurance_index - this->indexes_of_pattern_parts.at(pattern_index);
        if (idx >= 0 && idx < c.size()){
            c.at(idx) += 1;
        }
    }

    for (int i = 0; i < this->text.size() - this->joker_pattern.size() + 1; i++){
        if (c.at(i) == this->joker_pattern.size()){
            this->joker_pattern_occurence_indexes.push_back(i);
        }
    }
};

void Solver::printClassic(){
    for (auto pair : this->patterns_occurrence_indexes){
        int occurance_index = pair.first;
        int pattern_index = pair.second;
        std::cout << occurance_index + 1 << ' ' << pattern_index + 1 << std::endl;
    }
};

void Solver::printJoker(){
    for (int occurence_index: this->joker_pattern_occurence_indexes){
        std::cout << occurence_index + 1 << std::endl;
    }
};

void Solver::solve(){
    this->buildBohr();
    this->buildSuffixAndOutputLinks();

    switch (this->task_type){
        case Classic: 
            this->solveClassic();
            break;
        case Joker:
            this->solveJoker();
            break;
        default:
            this->solveClassic();
            break;
    }
    
};

void Solver::printAnswer(){
    switch (this->task_type){
        case Classic: 
            this->printClassic();
            break;
        case Joker:
            this->printJoker();
            break;
        default:
            this->printClassic();
            break;
    }
};


int main(){
    Solver *solver = new Solver(Joker);
    solver->solve();
    solver->printAnswer();
    return 0;
};
